hbook/h1d.go

```diff
@@ -186,6 +186,105 @@ func (h *H1D) Scale(factor float64) {
 	h.Binning.scaleW(factor)
 }
 
+// AddH1D returns the bin-by-bin summed histogram of h1 and h2
+// assuming their statistical uncertainties are uncorrelated.
+func AddH1D(h1, h2 *H1D) *H1D {
+
+	if h1.Len() != h2.Len() {
+		panic("hbook: h1 and h2 have different number of bins")
+	}
+
+	if h1.XMin() != h2.XMin() {
+		panic("hbook: h1 and h2 have different Xmin")
+	}
+
+	if h1.XMax() != h2.XMax() {
+		panic("hbook: h1 and h2 have different Xmax")
+	}
+
+	hsum := NewH1D(h1.Len(), h1.XMin(), h1.XMax())
+
+	for i := 0; i < hsum.Len(); i++ {
+		y := h1.Value(i) + h2.Value(i)
+		y1err2 := h1.Binning.Bins[i].SumW2()
+		y2err2 := h2.Binning.Bins[i].SumW2()
+		hsum.Binning.Bins[i].Dist.Dist.SumW = y
+		hsum.Binning.Bins[i].Dist.Dist.SumW2 = y1err2 + y2err2
+	}
+
+	// handle under/over flow
+	for i := range hsum.Binning.Outflows {
+		y := h1.Binning.Outflows[i].Dist.SumW + h2.Binning.Outflows[i].Dist.SumW
+		y1err2 := h1.Binning.Outflows[i].Dist.SumW2
+		y2err2 := h2.Binning.Outflows[i].Dist.SumW2
+		hsum.Binning.Outflows[i].Dist.SumW = y
+		hsum.Binning.Outflows[i].Dist.SumW2 = y1err2 + y2err2
+	}
+
+	return hsum
+}
+
+// AddScaledH1D returns the histogram with the bin-by-bin h1+alpha*h2
+// operation, assuming statistical uncertainties are uncorrelated.
+func AddScaledH1D(h1 *H1D, alpha float64, h2 *H1D) *H1D {
+	h2.Scale(alpha)
+	return AddH1D(h1, h2)
+}
+
+// DivH1D returns the bin-by-bin ratio histogram of h1 to h2
+// assuming their statistical uncertainties are uncorrelated.
+// Bins with NaN of +/- infinity value are set to 0, as well
+// as their uncertainty.
+func DivH1D(h1, h2 *H1D) *H1D {
+
+	if h1.Len() != h2.Len() {
+		panic("hbook: h1 and h2 have different number of bins")
+	}
+
+	if h1.XMin() != h2.XMin() {
+		panic("hbook: h1 and h2 have different Xmin")
+	}
+
+	if h1.XMax() != h2.XMax() {
+		panic("hbook: h1 and h2 have different Xmax")
+	}
+
+	hratio := NewH1D(h1.Len(), h1.XMin(), h1.XMax())
+
+	for i := 0; i < hratio.Len(); i++ {
```

Comment:
we have 4 of these loops.

perhaps (but this could be done in a later PR) we could factor this out into a helper function that takes the 2 slices of bins and the operation to apply?

Reply:
Yes this could be useful. One would need to pass two operations, one for the central value and one for the uncertainty propagation.

Reply:
couldn't this be done w/ one `func`?
ex:
```go
func h1dApply(dst, bins1, bins2 []Bin1D, fct func(b1, b2 Bin1D) (float64, float64)) {
    if len(bins1) != len(bins2) { ... }
    if len(bins1) != len(dst) { ... }
    for i := range dst {
        y, yerr := fct(bins1[i], bins2[i])
        dst[i].Dist.Dist.SumW = y
        dst[i].Dist.Dist.SumW2 = yerr
    }
}
```

Reply:
(BTW, we should probably also correctly handle the number of entries in the resulting `dst` histogram)

Reply:
haha, indeed a single function works! Let me try to implement it, this is easy (including the chec on total number of entries).

Reply:
Two short questions:
 1. this function would not take into account outflows, which are of a type `Dist` and not `Bin`. What would be the best here?
 2. for the handling of number of entries, I am not sure. When `fct` does `h1+h2`, this is straightforward, but what's about a ratio of histogram? Should I set `N` to the sum?

Reply:
ok please ignore the first question above.

Reply:
here's what ROOT does:
- https://root.cern/doc/v620/classTH1.html#ac782a09c31b4f7de40f8fd4f77efa090

a simple test yields:
```
root> auto h1 = new TH1D("h1","h1", 10, 0, 10)
root> h1->Fill(-0.5,1)
[...]
root> h1->Fill(5.9, 1)
(int) 6
root> h1->Fill(6.0, 0.5)
(int) 7

root> auto h2 = new TH1D("h2","h2", 10, 0, 10)
[...]
root> h1->GetEntries()
(double) 8.0000000
root> h2->GetEntries()
(double) 8.0000000

root> auto h3 = (*h1)+(*h2)
root> h3.GetEntries()
(double) 16.000000

root> auto h4 = (*h1)/(*h2)
root> h4.GetEntries()
(double) 2.1421582
```

YODA actually returns a scatter-2D:
```python
>>> import yoda
>>> h1 =yoda.Histo1D(6,0,6)
>>> h1.fill(1,1)
>>> h1.fill(2,1)
>>> h1.fill(3,1)
>>> h2 = yoda.Histo1D(6,0,6)
>>> h2.fill(1,2)
>>> h2.fill(2,2)
>>> h2.fill(3,2)
>>> h1/h2
<Scatter2D '/' 6 points>
>>> h1+h2
<Histo1D '/' 6 bins, sumw=9, xmean=2.00e+00>
```

I think it would actually be more sensible to go the YODA way.
what do you think?

Reply:
Indeed, I am up to return a `hbook.S2D` (thinking about it, I don't think the ratio of two histograms is an histogram). I'll implement the change.

So we miss only the handling of outflow using `h1dApply()` (or not), and we should be good.

Reply:
Heu ... I think I can remove the `DivH1D()` function :D https://godoc.org/go-hep.org/x/hep/hbook#S2D

![image](https://user-images.githubusercontent.com/30081646/79738442-09bdec80-82fd-11ea-8db5-68fc413a4603.png)

Reply:
ah! seems like I forgot I did cross my Ts at some point :P